### DISCUSSION TITLE
Linux distribution

### DISCUSSION THREAD

**Discussion by AlexandrePTJ** [AlexandrePTJ/kemai#100] [Polls]
AppImage is used to distribute kemai on Linux. It seems to be a good idea but several issues with AppImage / Qt / SSL / libc makes me reconsider this.

I do not want to support several packaging method, so I have to explore available solutions.

**Comment by virtulis** [comment]
Flatpak seems to be the least controversial "bundling" method overall, but of course the best solution is distro-specific packages (which don't have to be made by you!).

On that topic, I made an AUR "package" (unofficial script to build and install from upstream sources): https://aur.archlinux.org/packages/kemai-git (it's my first one!)

And one issue I had was that CMake really wanted to install header files for the "fetched" dependencies globally (which would conflict with actual packages providing those). Thankfully there is a flag to force it to FindPackage instead of Fetch, but I wonder if there's a better way to deal with this on the upstream side.

**Comment by AlexandrePTJ** [comment]
I will look at Flatpak, however I am considering providing a .deb for the last debian (or Ubuntu)
In my opinion, as long as software are opensource, it is better to be distro-specific.

@virtulis, thanks you for the AUR package :) Which flags are you talking about ?

**Reply by virtulis** [reply]
>  In my opinion, as long as software are opensource, it is better to be distro-specific.

I agree. While this is a "non-developer" piece of software so setting it up *should* be easy, from my experience as a user it doesn't really work very well anyway when packaging is done by the people who don't actually use the OS daily. 

Even for Debian/Ubuntu I'm not sure standalone `.deb` packages are a good idea, since that doesn't handle updates and might interfere with the updates of the OS itself due to outdated dependencies. I don't use Flatpak myself but as I understand the whole point of it is to isolate the package from the OS as much as possible.

> Which flags are you talking about ?

In [PKGBUILD](https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=kemai-git) I added these CMake flags:

```
    -DFETCHCONTENT_FULLY_DISCONNECTED=ON \
    -DFETCHCONTENT_TRY_FIND_PACKAGE_MODE=ALWAYS \
```

What this basically does is turn every `FetchContent_*` into `find_package` instead, which are then supplied from Arch repos and AUR via depends/makedepends lists. (I put range-v3 and magic_enum in makedepends since they seem to be header only)

Main reason I did that is because I wanted to use `cmake --install` to package exactly what *you* chose to install (because you probably know better than me, and might change it later). And CMake was very eager to install all those extra fetched packages as well, into `/usr/include` etc. There's probably a way to just exclude them from the install target but CMakeLists is a labyrinth to me (despite having dealt with it many times). And I do get it that it's easier to just fetch them instead of making the build instructions harder.

Now `spdlog` you should probably just tell the user to install from their repos. Especially since it looks dynamically linked.

**Reply by AlexandrePTJ** [reply]
In develop branch, I have added an option `KEMAI_BUILD_LOCAL_DEPENDENCIES` to decide if project will download and build its dependencies. Does it fit to what you expected ? @poelzi May be interested too.

**Reply by virtulis** [reply]
Sorry, I forgot to respond. Yes, it works great, thank you!
